### PR TITLE
feat: implement windowed porep

### PR DESCRIPTION
**Pull request by dignifiedquire**
- [x] construction
  - [x] replication
  - [x] decoding
  - [x] vanilla proving
  - [x] snark proving
- [x] integration
- [x] cleanup
- [x] v18 params
- [ ] ~~renaming~~

### CODE REVIEW

File: storage-proofs/src/stacked/proof.rs [hunk 1]
```diff
@@ -16,395 +18,603 @@ use crate::stacked::{
...
-                        .send(Message::Done)
-                        .expect("failed to finalize hasher");
+                if layer == layers {
+                    // on the last layer we encode the data
+                    let keyd = H::Domain::try_from_bytes(&key).unwrap();
+                    let data_node = H::Domain::try_from_bytes(
+                        &data_chunk[node * NODE_SIZE..(node + 1) * NODE_SIZE],
+                    )
+                    .unwrap();
+                    let encoded_node = decode::<H::Domain>(keyd, data_node);
```
Comment (porcuquine) [inline storage-proofs/src/stacked/proof.rs]: Let's call this `decoded_node`, for my sanity.

File: storage-proofs/src/stacked/proof.rs [hunk 2]
```diff
@@ -16,395 +18,603 @@ use crate::stacked::{
...
+                    ));
                 }
             }
 
-            // NOTE: this means we currently keep 2x sector size around, to improve speed.
             if let Some(ref mut exp_parents_data) = exp_parents_data {
                 exp_parents_data.copy_from_slice(&layer_labels);
             } else {
                 exp_parents_data = Some(layer_labels.clone());
             }
```
Comment (porcuquine) [inline storage-proofs/src/stacked/proof.rs]: I think this copying may be unnecessary on the last layer.

File: storage-proofs/src/stacked/proof.rs [hunk 3]
```diff
@@ -16,395 +18,603 @@ use crate::stacked::{
...
+
+        // TODO: make window and wrapper challengs have different number of challenges
+        let window_challenges =
+            pub_inputs.all_challenges(layer_challenges, window_graph.size(), Some(k));
+
+        let window_proofs: Vec<_> = window_challenges
+            .into_par_iter()
+            .enumerate()
+            .map(|(challenge_index, challenge)| {
+                let window_index = get_window_index(challenge);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: fixed

File: storage-proofs/src/stacked/proof.rs [hunk 4]
```diff
@@ -16,395 +18,603 @@ use crate::stacked::{
...
+
+        trace!(" challenge {} ({})", challenge, challenge_index);
+        assert!(challenge < window_graph.size(), "Invalid challenge");
+        assert!(challenge > 0, "Invalid challenge");
+
+        let layers = layer_challenges.layers();
 
-        let graph_size = graph.size();
+        // Initial data layer openings (c_X in Comm_D)
+        let comm_d_proof = MerkleProof::new_from_proof(&t_aux.tree_d.gen_proof(challenge));
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: fixed

File: storage-proofs/src/stacked/proof.rs (hunk 2 above)
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: I believe that is correct

Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: fixed

File: storage-proofs/src/stacked/proof.rs (hunk 1 above)
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: fixed

File: storage-proofs/src/stacked/proof.rs [hunk 5]
```diff
@@ -97,14 +93,13 @@ impl<'a, H: 'static + Hasher, G: 'static + Hasher> StackedDrg<'a, H, G> {
             pub_inputs.all_challenges(layer_challenges, window_graph.size(), Some(k));
 
         let window_proofs: Vec<_> = window_challenges
-            .into_par_iter()
+            // .into_par_iter()
```
Comment (porcuquine) [inline storage-proofs/src/stacked/proof.rs]: What did you intend here?

Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: that was just for debugging

File: storage-proofs/src/stacked/porep.rs [hunk 6]
```diff
@@ -39,23 +34,31 @@ impl<'a, 'c, H: 'static + Hasher, G: 'static + Hasher> PoRep<'a, H, G> for Stack
...
+        let mut window = data[window_start..window_end].to_vec();
+
+        Self::extract_single_window(pp, replica_id, &mut window, window_start_index);
+
+        let node_window_index = node % pp.window_size_nodes();
+        let start = node_window_index * NODE_SIZE;
+        let end = (node_window_index + 1) * NODE_SIZE;
+        let node = window[start..end].to_vec();
+
+        Ok(node)
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/porep.rs]: agreed

File: storage-proofs/src/stacked/params.rs [hunk 7]
```diff
@@ -155,41 +176,115 @@ pub struct PrivateInputs<H: Hasher, G: Hasher> {
...
+    pub labeling_proof: LabelingProof<H>,
 }
 
-impl<H: Hasher, G: Hasher> Proof<H, G> {
+impl<H: Hasher> WrapperProof<H> {
     pub fn comm_r_last(&self) -> &H::Domain {
         self.comm_r_last_proof.root()
     }
 
+    /// Verify the full proof.
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: not really, but I can see how it is confusing

File: storage-proofs/src/stacked/params.rs [hunk 8]
```diff
@@ -198,80 +293,111 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
...
-        graph.parents(challenge, &mut parents);
-        check!(self.replica_column_proofs.verify(challenge, &parents));
-
-        check!(self.verify_final_replica_layer(challenge));
+        check!(self
+            .replica_column_proof
+            .verify(challenge, &pub_params.window_graph, comm_c));
 
-        check!(self.verify_labels(replica_id, &pub_params.layer_challenges, challenge_index));
+        check!(self.verify_labels(replica_id, pub_params.config.layers()));
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/params.rs]: I do, take a look at how `verify_labels` pulls the data directly from `replica_column_proof.c_x`

File: storage-proofs/src/stacked/proof.rs [hunk 9]
```diff
@@ -31,425 +34,626 @@ pub struct StackedDrg<'a, H: 'a + Hasher, G: 'a + Hasher> {
...
+                            parents_data.clone(),
+                        );
+
+                        {
+                            let expected_labeled_node = replica_column_proof
+                                .c_x
+                                .get_node_at_layer(window_index, layer);
+
+                            assert!(
+                                proof.verify(&pub_inputs.replica_id, &expected_labeled_node),
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: what do you mean with "correct parrents" check here?

File: storage-proofs/src/stacked/proof.rs [hunk 10]
```diff
@@ -31,425 +34,626 @@ pub struct StackedDrg<'a, H: 'a + Hasher, G: 'a + Hasher> {
...
-                                data_at_node(parents_data, *parent as usize).expect("invalid node");
-                            hasher.input(&buf);
-                        }
-                    }
-                }
+        // hash replica id
+        base_hasher.input(AsRef::<[u8]>::as_ref(replica_id));
+
+        for layer in 1..=layers {
+            trace!("generating layer: {}", layer);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: There are unfortunately some differences, making this hard. a large part is already shared through `create_key` though

File: storage-proofs/src/stacked/proof.rs [hunk 11]
```diff
@@ -31,425 +34,626 @@ pub struct StackedDrg<'a, H: 'a + Hasher, G: 'a + Hasher> {
...
-            },
-            column_hashes,
+            LabelsCache::<H>::from_stores(labels),
+            Labels::<H>::new(configs),
         ))
     }
 
+    #[allow(clippy::too_many_arguments)]
+    fn encode_window_layer(
+        layer: usize,
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: hmmm `label_and_encode_window_layer`?

File: storage-proofs/src/stacked/proof.rs [hunk 12]
```diff
@@ -31,425 +34,626 @@ pub struct StackedDrg<'a, H: 'a + Hasher, G: 'a + Hasher> {
...
+                        data_chunk,
+                        window_index,
+                    );
+
+                    if layer < layers {
+                        if let Some(ref mut exp_parents_data) = exp_parents_data {
+                            exp_parents_data.copy_from_slice(&layer_labels);
+                        } else {
+                            exp_parents_data = Some(layer_labels.clone());
+                        }
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: this else branch happens on the first iteration, when `exp_parent_data` is `None`

File: storage-proofs/src/stacked/params.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/stacked/params.rs]: I see it now, thanks.

File: storage-proofs/src/circuit/stacked/params.rs [hunk 13]
```diff
@@ -58,66 +71,147 @@ impl<H: Hasher, G: Hasher> Proof<H, G> {
...
+        replica_id: &[Boolean],
+    ) -> Result<(), SynthesisError> {
+        let WrapperProof {
+            comm_r_last_proof,
+            comm_q_parents_proofs,
+            labeling_proof,
+        } = self;
 
         let comm_r_last_data_leaf =
             comm_r_last_proof.alloc_value(cs.namespace(|| "comm_r_last_data_leaf"))?;
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/params.rs]: we do, I forgot about that

File: storage-proofs/src/stacked/proof.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/stacked/proof.rs]: I was confused about how some of the proving and verification fit together. This was a note to myself trying to  puzzle it through. Given your comments above, I think the necessary checks are happening. I think the way the parts are decomposed is a little confusing, but this was also just me trying to get through a long review.

File: storage-proofs/src/stacked/proof.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/stacked/proof.rs]: I want a name making it obvious that we don't encode most layers. I think leaving `encode` out completely would be okay, with a comment noting that the last layer is also encoded as an optimization. If you want to be hyper-verbose, then I guess `label_and_optionally_encode_window_layer` would be accurate (but I don't love it even a little).

File: storage-proofs/src/stacked/proof.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/stacked/proof.rs]: Right, thanks. My eyes got tripped up on the nested `if`.

File: filecoin-proofs/src/constants.rs [hunk 14]
```diff
@@ -13,6 +15,14 @@ pub const SECTOR_SIZE_256_MIB: u64 = 1 << 28;
 
 pub const SECTOR_SIZE_1_GIB: u64 = 1 << 30;
 
+pub const WINDOW_SIZE_NODES_ONE_KIB: usize = 512 / NODE_SIZE;
+
+pub const WINDOW_SIZE_NODES_16_MIB: usize = (4 * 1024 * 1024) / NODE_SIZE;
+
+pub const WINDOW_SIZE_NODES_256_MIB: usize = (8 * 1024 * 1024) / NODE_SIZE;
+
+pub const WINDOW_SIZE_NODES_1_GIB: usize = (64 * 1024 * 1024) / NODE_SIZE;
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply filecoin-proofs/src/constants.rs]: added some comments

File: storage-proofs/src/stacked/porep.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/porep.rs]: improved, there is now an `extract_range` method

File: storage-proofs/src/circuit/stacked/params.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/circuit/stacked/params.rs]: fixed

File: storage-proofs/src/stacked/proof.rs [hunk 15]
```diff
@@ -716,15 +1071,115 @@ mod tests {
...
+            println!("decoding {} - {}", start, len);
+            let nodes = StackedDrg::<H, Blake2sHasher>::extract_range(
+                &pp,
+                &replica_id,
+                &data_copy,
+                Some(config.clone()),
+                *start * NODE_SIZE,
+                *len * NODE_SIZE,
+            )
+            .expect("failed to extract data");
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: correct, there were even more bugs. all squashed now

File: storage-proofs/src/stacked/proof.rs [hunk 16]
```diff
@@ -487,50 +490,33 @@ impl<'a, H: 'static + Hasher, G: 'static + Hasher> StackedDrg<'a, H, G> {
...
-        let last_window_len = if (num_bytes + first_window_offset) % pp.window_size_bytes() > 0 {
-            (num_bytes + first_window_offset) % pp.window_size_bytes()
-        } else {
-            pp.window_size_bytes()
-        };
-
         // determine the number of windows need to be decoded
-        let num_windows = (num_bytes as f64 / pp.window_size_bytes() as f64).ceil() as usize;
+        let num_windows =
+            ((offset + num_bytes) as f64 / pp.window_size_bytes() as f64).ceil() as usize;
```
Comment (porcuquine) [inline storage-proofs/src/stacked/proof.rs]: What if `num_bytes` is very small and `offset` is large. Won't `num_windows` be large, when it should be `1`?

I think what you are calling `num_windows` is actually `last_window_index`, and `num_window = last_window_index - first_window_index` or something like that.

Comment (dignifiedquire) [reply storage-proofs/src/stacked/proof.rs]: I need to update the comment, and maybe rename the variable. If you look how I use it, I first take `num_windows` and then skip `first_window_index`. It now is more like `last_window_index`

Comment (porcuquine) [reply storage-proofs/src/stacked/proof.rs]: Got it. I do recommend making those clarification so correctness is obvious to future readers. But looks like this is in order then.

**Pull request merged**